server/task_scheduler.go

```diff
@@ -119,21 +119,53 @@ func (s *TaskScheduler) Run(ctx context.Context, wg *sync.WaitGroup) {
 						continue
 					}
 
-					tasks.mu.Lock()
-					if _, ok := tasks.running[task.ID]; ok {
-						tasks.mu.Unlock()
+					// loaded means the task is already running so we continue
+					if _, loaded := s.runningTask.LoadOrStore(task.ID, &api.Progress{CreatedTs: time.Now().Unix()}); loaded {
 						continue
 					}
-					tasks.running[task.ID] = true
-					tasks.mu.Unlock()
 
 					go func(task *api.Task) {
 						defer func() {
-							tasks.mu.Lock()
-							delete(tasks.running, task.ID)
-							tasks.mu.Unlock()
+							s.runningTask.Delete(task.ID)
+						}()
+						taskComplete := make(chan struct{})
+						go func() {
+							snapshotProgress := func() {
+								value, ok := s.runningTask.Load(task.ID)
+								if !ok {
+									log.Error("Failed to load task from task scheduler", zap.Int("id", task.ID))
+									return
+								}
+								progress, ok := value.(*api.Progress)
+								if !ok {
+									log.Error("Failed to assert the type to be progress", zap.Int("id", task.ID), zap.Any("progress", progress))
+									return
+								}
+								task.Progress.Lock()
+								progress.Lock()
+								progress.TotalUnit = task.Progress.TotalUnit
+								progress.CompletedUnit = task.Progress.CompletedUnit
+								progress.UpdatedTs = time.Now().Unix()
```

Comment:
We should copy the whole progress as is because the task executor is the only source of updates to progress. The map in the scheduler is just a copied snapshot.